google_authorized_keys/main_test.go

```diff
@@ -0,0 +1,232 @@
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+)
+
+func stringSliceEqual(a, b []string) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i, v := range a {
+		if v != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func boolToStr(b *bool) string {
+	if b == nil {
+		return "<nil>"
+	}
+	return strconv.FormatBool(*b)
+}
+
+var t = true
+var f = false
+var truebool *bool = &t
+var falsebool *bool = &f
+
+func TestParseSSHKeys(t *testing.T) {
+	keys := []string{
+		"# Here is some random data in the file.",
+		"usera:ssh-rsa AAAA1234USERA",
+		"userb:ssh-rsa AAAA1234USERB",
+		`usera:ssh-rsa AAAA1234 google-ssh {"userName":"usera@example.com","expireOn":"2095-04-23T12:34:56+0000"}`,
+		`usera:ssh-rsa AAAA1234 google-ssh {"userName":"usera@example.com","expireOn":"2020-04-23T12:34:56+0000"}`,
+	}
+	expected := []string{
+		"ssh-rsa AAAA1234USERA",
+		`ssh-rsa AAAA1234 google-ssh {"userName":"usera@example.com","expireOn":"2095-04-23T12:34:56+0000"}`,
+	}
+
+	user := "usera"
+
+	if got, want := parseSSHKeys(user, keys), expected; !stringSliceEqual(got, want) {
+		t.Errorf("ParseSSHKeys(%s,%s) incorrect return: got %v, want %v", user, keys, got, want)
+	}
+
+}
+
+func TestCheckWinSSHEnabled(t *testing.T) {
+	tests := []struct {
+		instanceEnable *bool
+		projectEnable  *bool
+		expected       bool
+	}{
+		{
+			instanceEnable: truebool,
+			projectEnable:  nil,
+			expected:       true,
+		},
+		{
+			instanceEnable: falsebool,
+			projectEnable:  nil,
+			expected:       false,
+		},
+		{
+			instanceEnable: falsebool,
+			projectEnable:  truebool,
+			expected:       false,
+		},
+		{
+			instanceEnable: nil,
+			projectEnable:  truebool,
+			expected:       true,
+		},
+		{
+			instanceEnable: nil,
+			projectEnable:  falsebool,
+			expected:       false,
+		},
+		{
+			instanceEnable: nil,
+			projectEnable:  nil,
+			expected:       false,
+		},
+	}
+	for _, tt := range tests {
+		instanceAttributes := attributes{EnableWindowsSSH: tt.instanceEnable}
+		projectAttributes := attributes{EnableWindowsSSH: tt.projectEnable}
+		if got, want := checkWinSSHEnabled(&instanceAttributes, &projectAttributes), tt.expected; got != want {
+			t.Errorf("checkWinSSHEnabled(%s, %s) incorrect return: got %v, want %v", boolToStr(tt.instanceEnable), boolToStr(tt.projectEnable), got, want)
```

Comment:
As discussed, the boolToStr isn't really necessary, but just a way to make the test output more useful since the inputs to the function are pointers to boolean values and %v will just print the address and using * to dereference will give an error if it's nil.